recipes/lora_finetune_distributed.py

```diff
@@ -307,16 +306,36 @@ def _setup_model(
                 base_model_state_dict_keys=base_model_state_dict.keys(),
             )
 
-            # Load both the base model weights and (if available) the adapter weights. Both
-            # of this should happen only on Rank 0
-            model.load_state_dict(base_model_state_dict, strict=False)
-            if lora_weights_state_dict:
-                model.load_state_dict(lora_weights_state_dict, strict=False)
+        self.adapter_params = get_adapter_params(model)
+        set_trainable_params(model, self.adapter_params)
 
-        else:
-            # For non-zero ranks, load the model on meta device
-            with utils.set_default_dtype(self._dtype), torch.device("meta"):
-                model = config.instantiate(cfg_model)
+        if enable_activation_checkpointing:
+            utils.set_activation_checkpointing(
+                model, auto_wrap_policy={modules.TransformerDecoderLayer}
+            )
+
+        for m in model.modules():
+            if isinstance(m, modules.TransformerDecoderLayer):
+                fully_shard(m)
+        fully_shard(model)
+
+        utils.load_from_full_model_state_dict(
+            model, base_model_state_dict, self._device, self._is_rank_zero
+        )
+        if lora_weights_state_dict:
+            utils.load_from_full_model_state_dict(
+                model, lora_weights_state_dict, self._device, self._is_rank_zero
+            )
+
```

Comment:
pros and cons of meta init. pros is 4.5x speed up during model init and thus shorter TTFB. cons is user need to call `initialize_parameters` on `LoRALinear` explicitly to move them from `meta` to `gpu`

Reply:
you are right. when finetuning from a original HF checkpoint, `lora_weights_state_dict = None`

for resuming training, `lora_weights_state_dict is not None` and we avoided calling `m.initialize_parameters()` again

Reply:
Got you, thanks so much for the explanation! I think something that would be super helpful would be document here in the form of comments the relationship between:
- the modules on which we call ```fully_shard```
- init on meta device
- calling ```initialize_parameters``` and ```reset_parameters```

Also I think there was a technical reason with FSDP1 to call the function ```reset_parameters```. Is that still true? Or can we standardize this with ```initialize_parameters``` in the  modules code? Happy to chat about this offline!

Reply:
good point! will add comment to explain `fully_shard`, meta init, and `reset/initialize_parameters`

FSDP1 calls `reset_parameters` for the exact same reason FSDP2 call `reset/initialize_parameters`: RoPE are not covered in checkpoints, `lora_a` and `lora_b` are not covered in checkpoints for `resume_training=False`

It's just FSDP1 have a contract to call overrided `nn.Module.reset_parameter` through `FSDP(model, param_init=)`, but FSDP2 does not impose overriding `reset_parameter`. now use can name it `reset_parameter` or `initialize_parameters`